undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentProcessor.java

```diff
@@ -129,12 +129,43 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
         if (warMetaData == null) {
             return;
         }
-        String hostName = hostNameOfDeployment(warMetaData, defaultHost);
-        processDeployment(warMetaData, deploymentUnit, phaseContext.getServiceTarget(), hostName);
+        String deploymentName;
+        if (deploymentUnit.getParent() == null) {
+            deploymentName = deploymentUnit.getName();
+        } else {
+            deploymentName = deploymentUnit.getParent().getName() + "." + deploymentUnit.getName();
+        }
+        final UndertowService undertowService = (UndertowService) phaseContext.getServiceRegistry().getService(UndertowService.UNDERTOW).getValue();
+        final Set<Server> servers = undertowService.getServers();
```

Comment:
"I really don't see any scenario where this would work wrongly, as servers & hosts are always started before any deployment can kick in. "

I'd love to be wrong about this, but no, they are not. Not unless there is a dependency relationship. The add/enable ops for deployments are just another step during boot and the DeploymentUnitPhaseServices that invoke DUPs in start are just services. If a DUPS starts in boot after the relevant server and host services without a dependency relationship, it's luck. Maybe typically expected luck but still luck.